crates/constraint-vm/src/alu.rs

```diff
@@ -22,6 +25,33 @@ pub(crate) fn mod_(a: Word, b: Word) -> OpResult<Word> {
     a.checked_rem(b).ok_or(AluError::DivideByZero.into())
 }
 
+pub(crate) fn shl(a: Word, b: Word) -> OpResult<Word> {
+    check_shift_bounds(b)?;
+    Ok(a << b)
+}
+
+pub(crate) fn shr(a: Word, b: Word) -> OpResult<Word> {
+    check_shift_bounds(b)?;
+    // casts are safe and turn this into a logical shift
+    Ok(((a as u64) >> b) as i64)
+}
+
+pub(crate) fn arithmetic_shr(a: Word, b: Word) -> OpResult<Word> {
+    check_shift_bounds(b)?;
+    Ok(a >> b)
+}
+
+const BITS_IN_WORD: Word = core::mem::size_of::<Word>() as Word * 8;
+
+#[inline]
+fn check_shift_bounds(b: Word) -> OpResult<()> {
+    let bounds = 0..BITS_IN_WORD;
+    if !bounds.contains(&b) {
+        return Err(AluError::Overflow.into());
+    }
+    Ok(())
+}
```

Comment:
Another option might be to use `.checked_shl/shr` which I think does this for us

Reply:
Yeh I tried that but they all take other datatypes on the rhs like u32. This forces you to do a try into then also call checked. Where as this way there's just a single check